### ISSUE TITLE
The web cmdlets (`Invoke-WebRequest` and `Invoke-RestMethod`) shouldn't use a user-agent string that suggests an interactive browser; instead, they should suggest a download utility such as `curl`

### ISSUE THREAD

**Issue by mklement0** [PowerShell/PowerShell#24425]
### Summary of the new feature / enhancement

Many web servers modify their behavior based on the user-agent string provided by the client:

* (a) If the user-agent string suggests a client that is a _full-fledged, interactive browser_, an HTML response with embedded JavaScript may be returned - and in such clients that usually results in useful _interactive_ behavior, such as popping up a download-to-a-local-file dialog.

* (b) Otherwise, i.e. if the user-agent string suggests a _download utility_ such as `curl`, they may issue a redirect to a download target or provide a JSON response that can be acted upon _programmatically_.

Unfortunately, PowerShell's web cmdlets currently default to a user-agent string suggestive of (a), which results in unhelpful behavior in many cases - see [this Stack Overflow post](https://stackoverflow.com/q/79076263/45375) for an example.

Defaulting to a user-agent string suggestive of (b) is therefore a more sensible choice, in the simplest form `curl`, based on what the widely used `curl` utility uses; specifically, it  uses `curl/<version-number>`, e.g. `curl/8.9.1`, and provides _no_ information about the caller's host platform.

For now, using something like `-UserAgent curl` is required as a workaround.

### Proposed technical implementation details (optional)

* I understand that such a change would technically a breaking one, and its impact on existing code would have to be assessed.

* At the very least, the problem should be _documented_.

**Comment by 237dmitry** [comment]
Just out of curiosity. What problems? In my experience, curl sometimes doesn't work, but IWR does. And vice versa. However, I can't give any specific examples. It probably sounds implausible.

**Comment by mklement0** [comment]
An example is the following URL (taken from the linked Stack Overflow post): https://downloads.sourceforge.net/project/boost/boost-binaries/1.74.0/boost_1_74_0-msvc-14.2-64.exe is a _direct download_ when you target it with `curl`, but with the web cmdlets, due to sending a user-agent string that is suggestive of an interactive browser - you get an _HTML page_ back (which in true browser would trigger a download dialog via JavaScript).

In general, the web cmdlets should identify themselves as the _automated tools_ that they are _by default_ (you can always override it with `-UserAgent`) rather than pretend to be _interactive browsers_.

**Comment by 237dmitry** [comment]
![ArcoLinux_2024-10-14_19-36-37](https://github.com/user-attachments/assets/211f4a34-e3f1-49a3-aa60-874e2e14319c)

**Comment by mklement0** [comment]
Indeed: what was download was an _HTML page_, not an `.exe` file, which you can easily verify by opening the output file in a text editor. The true file size should be `170459429` bytes (ca. 163 MB), which is what you get with: 

    curl -L -o foo.exe https://downloads.sourceforge.net/project/boost/boost-binaries/1.74.0/boost_1_74_0-msvc-14.2-64.exe

 or - by making `Invoke-WebRequest` pose as `curl` - with:

    Invoke-WebRequest -UserAgent curl -OutFile foo.exe https://downloads.sourceforge.net/project/boost/boost-binaries/1.74.0/boost_1_74_0-msvc-14.2-64.exe

The point of this request is that you shouldn't have to use `-UserAgent curl` - a user-agent string like that should be the default.

**Comment by 237dmitry** [comment]
Yes, indeed 162Mb and 76Kb are not the same

**Comment by iSazonov** [comment]
Current `user-agent` contains both `Mozilla/5.0` and `PowerShell/7.5.0`. So any web server could be smart enough to return a customized result. The main thing here is that we explicitly inform the server that the request comes from PowerShell.

**Comment by rhubarb-geek-nz** [comment]
> The point of this enhancement request is that you shouldn't have to use `-UserAgent curl` - a user-agent string like that should be the default.

Have you tried different values for UserAgent to see the effect? If curl is treated differently to a web browser is it because the server has explicit checks for "curl". If Invoke-WebRequest said user agent PowerShell, why should a web-server not think that is simply another type of browser?

Have you tried changing the "Accept" value to say you want a zip or exe, and not an html?

**Comment by rhubarb-geek-nz** [comment]
> * At the very least, the problem with the current behavior should be _documented_.

Given that there is no way to predict, manage or measure the impact of such a change, I suggest that documentation is the safest solution.

**Comment by mklement0** [comment]
> Have you tried changing the "Accept" value to say you want a zip or exe, and not an html?

`-UserAgent curl` is a simpler workaround, but, as noted, the very purpose of this enhancement request is not to need a workaround.

But it is a fair point that assessing the impact of the proposed enhancement is very challenging in this case.

> If Invoke-WebRequest said user agent PowerShell, why should a web-server not think that is simply another type of browser?

It is the legacy `Mozilla/5.0` part in the web cmdlets' user-agent string that makes the target server think the client is an interactive browser, as all major browsers start their user-agent strings with this text.

https://useragentstring.com/pages/useragentstring.php has sample user-agent strings, categorized by client type (browser, crawler)

Whether it is typical for servers to check for `curl` explicitly I cannot say, but given `curl`'s ubiquity, it is safe to assume that replacing `Mozilla/5.0` with, e.g. `curl/8.7.1` in the current user-agent string would do the trick.

That is (and I should have made that clearer that I wasn't looking to _remove_ the `PowerShell/7.5.0` part, @iSazonov), something like the following would work (it does with https://downloads.sourceforge.net/project/boost/boost-binaries/1.74.0/boost_1_74_0-msvc-14.2-64.exe):

```powershell
curl/8.7.1 (Macintosh; Darwin 23.6.0 Darwin Kernel Version 23.6.0: Mon Jul 29 21:14:30 PDT 2024; root:xnu-10063.141.2~1/RELEASE_ARM64_T6000; en-US) PowerShell/7.5.0
```

Using _only_ the PowerShell part, at the start, works too with the URL at hand:

```powershell
PowerShell/7.5.0 (Macintosh; Darwin 23.6.0 Darwin Kernel Version 23.6.0: Mon Jul 29 21:14:30 PDT 2024; root:xnu-10063.141.2~1/RELEASE_ARM64_T6000; en-US)
```

However, the user-agent string-analyzing API at https://useragentstring.com/pages/api.php recognizes only the `curl`-prefixed version:

```powershell
Invoke-RestMethod (
  'https://useragentstring.com?uas={0}&getJSON=all' -f 
  'curl/8.7.1 (Macintosh; Darwin 23.6.0 Darwin Kernel Version 23.6.0: Mon Jul 29 21:14:30 PDT 2024; root:xnu-10063.141.2~1/RELEASE_ARM64_T6000; en-US) PowerShell/7.5.0'
)
```

Output:
```none
agent_type        : Librarie
agent_name        : curl
agent_version     : 8.7.1
os_type           : Darwin
os_name           : Darwin
os_versionName    : 
os_versionNumber  : 23.6.0
os_producer       : 
os_producerURL    : 
linux_distibution : Null
agent_language    : English - United States
agent_languageTag : en-US
```

**Comment by 237dmitry** [comment]
How did you find `curl` value for `UserAgent` parameter? It is not in known PowerShell user agents:

```powershell
#!/usr/bin/env -S pwsh -nop

$uas = ([Microsoft.PowerShell.Commands.PSUserAgent] |
        Get-Member -Static -MemberType Property).Name

$np = [Microsoft.PowerShell.Commands.PSUserAgent].GetMembers('Static, NonPublic').Where(
          { $_.Name -eq 'UserAgent' }
      ).GetValue($null,$null)

foreach ($ua in $uas)
{
    [pscustomobject] @{
        UserAgent = $ua
        Value = [Microsoft.PowerShell.Commands.PSUserAgent]::$ua
    }
}

[pscustomobject] @{ UserAgent = "NonPublic"; Value = $np }
```


```
UserAgent        Value
---------        -----
Chrome           Mozilla/5.0 (Linux; ArcoLinux; ru-RU) AppleWebKit/534.6 (KHTML, like Gecko) Chrome/7.0.500.0 Safari/534.6
FireFox          Mozilla/5.0 (Linux; ArcoLinux; ru-RU) Gecko/20100401 Firefox/4.0
InternetExplorer Mozilla/5.0 (compatible; MSIE 9.0; Linux; ArcoLinux; ru-RU)
Opera            Opera/9.70 (Linux; ArcoLinux; ru-RU) Presto/2.2.1
Safari           Mozilla/5.0 (Linux; ArcoLinux; ru-RU) AppleWebKit/533.16 (KHTML, like Gecko) Version/5.0 Safari/533.16
NonPublic        Mozilla/5.0 (Linux; ArcoLinux; ru-RU) PowerShell/7.5.0
```

**Comment by mklement0** [comment]
@237dmitry: I didn't _find_ it, because it isn't there. The suggestion is to _introduce it_, and, above all, to use it _by default_.

**Comment by 237dmitry** [comment]
> The suggestion is to introduce it

But it works already... I just want to understand how did you find this value?

**Comment by rhubarb-geek-nz** [comment]
> The suggestion is to _introduce it_, and, above all, to use it _by default_.

The principle of changing defaults just because they are inappropriate "in many cases" is a very bad idea when dealing with scripting systems. Scripts exist that use and rely on default behaviour, changing those defaults may break those scripts. With automated updates users have no control over which version they have, and with a monolithic system like PowerShell no practical way of tracking and managing these minor changes. Likewise the PowerShell team has absolutely no way of assessing the impact.

**Comment by mklement0** [comment]
You may have missed it, because I only edited it into my previous comment later:

> That said, it's hard to imagine that the current default behavior is relied upon, given that receiving an HTML response when _programmatic processing_ is intended is unlikely to be helpful, given that PowerShell 7 not only lacks HTML parsing abilities [out of the box], but due to the fact that the data of interest may additionally only be revealed through _client-side JavaScript execution_.

In other words: To me, the proposed enhancement is a candidate for a [bucket 3](https://github.com/PowerShell/PowerShell/blob/master/docs/dev-process/breaking-change-contract.md#bucket-3-unlikely-grey-area) breaking change: one that provides tangible benefits while being _unlikely_ to break existing code.

@237dmitry: I don't understand what you're saying.

**Comment by 237dmitry** [comment]
> I don't understand what you're saying.

Ok. You used `curl` value for `UserAgent` parameter and it works. My question is how did you find this value, you didn't make it up and it's not in the list of valid `UserAgent` parameter values.

**Comment by 237dmitry** [comment]
I tried and it works too:

```
Invoke-WebRequest -UserAgent aaaaa -out foo.exe https://downloads.sourceforge.net/project/boost/boost-binaries/1.74.0/boost_1_74_0-msvc-14.2-64.exe
```

So `curl` is not a workaround, rather any string is.

**Comment by mklement0** [comment]
Yes, it is the _absence of `Mozilla/5.0`_ that makes the difference in this case.

And analogous to this part of the string currently being included  _for compatibility_ in order to signal to user-agent sniffing serves that the client is a full-fledged browser, my suggestion is to instead use something like `curl/8.7.1` to signal that the web cmdlets are _curl-like_ in their purpose (non-interactive, no JavaScript capabilities, meant for interacting with web _APIs_); and choosing `curl` as the reference point, as the probably most widely used client in that space, make sense to me.

(As for how I determined that string: look at the output from `curl https://postman-echo.com/get`)

[My previous comment](https://github.com/PowerShell/PowerShell/issues/24425#issuecomment-2414533466) also showed an alternative to incorporating `curl`, which basically moves the `PowerShell/<version>` part to the front, replacing `Mozilla/5.0`.

At the end of the day, the gist of this enhancement is to make user-agent-sniffing servers _not_ think that the web cmdlets are full-fledged browsers, and for that, removing `Mozilla/5.0` alone may be sufficient - but the additional presence of `curl/<version>` wouldn't hurt either (though it does require settling on a specific, arbitrary version), though I don't know how common it is for servers to look for this substring.

**Comment by rhubarb-geek-nz** [comment]
> That said, it's hard to imagine that the current default behavior is relied upon
...
> not only lacks HTML parsing abilities [out of the box]

Is that where we are? We can only imagine and consider scenarios in a closed little world?

I take the view that defaults whether documented or not are part of the existing behaviour and we have no way of assessing the impact of changing those. The work around is simple, predictable and easily understood.

**Comment by mklement0** [comment]
Is that where we are? Repeatedly making cantankerous, borderline disrespectful statements instead of engaging with arguments?

**Comment by rhubarb-geek-nz** [comment]
> Is that where we are? Repeatedly making cantankerous, borderline disrespectful statements instead of engaging with arguments?

My understanding of PowerShell is it is designed for integration, so you could have any processing you like following from the output of Invoke-WebRequest. We are not limited to what comes in the box. For example it could be used for website testing where it validates signatures of pages without having to have deep knowledge. It could pass the pages to html or xml parsers for further processing. In that scenerio it would be depending on the user agent identifying itself as a browser.
